src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -223,64 +231,193 @@ private UUID getParticipantEntitySetId( UUID studyId ) {
 
-        ParticipationStatus status = getParticipationStatus( studyId, participantId );
-        if ( ParticipationStatus.NOT_ENROLLED.equals( status ) ) {
-            logger.warn( "participantId = {} is not enrolled, ignoring data upload", participantId );
-            return 0;
+        return reserveEntityKeyId(
+                recordedByEntitySetId,
+                ImmutableList.of( dateLoggedPropertyTypeId, stringIdPropertyTypeId, appPackageNamePropertyTypeId ),
+                data,
+                dataIntegrationApi
+        );
+    }
+
+    private UUID reserveUserAppEntityKeyId(
+            Map<UUID, Set<Object>> entityData,
+            DataIntegrationApi dataIntegrationApi ) {
+
+        return reserveEntityKeyId(
+                userAppsEntitySetId,
+                ImmutableList.of( appPackageNamePropertyTypeId ),
+                entityData,
+                dataIntegrationApi
+        );
+    }
+
+    private void createUserAppsEntitiesAndAssociations(
+            DataApi dataApi,
+            DataIntegrationApi dataIntegrationApi,
+            List<SetMultimap<UUID, Object>> data,
+            UUID deviceEntityKeyId,
+            UUID participantEntitySetId,
+            UUID participantEntityKeyId,
+            String participantId,
+            String deviceId ) {
+
+        /*
+         * Most of the data pushed by devices does not correspond to apps that were visible in the UI.
+         * Ideally we should only record UsageStats for apps that the user actually used
+         * Method getTotalTimeVisible() available in API level 29 would be the ideal filter since it gives the
+         * total time a package's activity was visible in the UI. However, our android app currently supports minimum API
+         * level 21. so the next best option would be getTotalTimeInForeground() which gives the total time spent in the
+         * foreground, measured in milliseconds
+         *
+         * https://developer.android.com/reference/android/app/usage/UsageStats
+         */
+        List<SetMultimap<UUID, Object>> userAppsData = data
+                .stream()
+                .filter( entry -> entry.containsKey( durationPropertyTypeId ) &&
+                        entry.containsKey( startDateTimePropertyId ) &&
+                        entry.containsKey( appNamePropertyTypeId ) &&
+                        Long.parseLong( entry.get( durationPropertyTypeId ).iterator().next().toString() ) > 0 )
+                .collect( Collectors.toList() );
+
+        for ( int i = 0; i < userAppsData.size(); ++i ) {
+
+            Set<DataEdgeKey> dataEdgeKeys = new HashSet<>();
+            String appPackageName = userAppsData.get( i ).get( appPackageNamePropertyTypeId ).iterator().next()
+                    .toString();
+            String dateLogged = parseDateTime( userAppsData.get( i ).get( startDateTimePropertyId ).iterator().next()
+                    .toString() );
```

Comment:
I see the `entity.get(property).iterator().next().toString()` in quite a lot of places. I don't have a recommendation for an alternative but I worry about this being problematic, i.e. exceptions being thrown.

Reply:
It's probably not a good idea to assume that all the data coming from the devices is ok (has no null values), so maybe a `try catch` wrapper would be needed.